nf_core/modules.py

```diff
@@ -231,3 +237,481 @@ def has_valid_pipeline(self):
+            self.passed.append("options specified in {}".format(file))
+        else:
+            self.warned.append("options not specified in {}".format(file))
+
+        # Go through module main.nf file and switch state according to current section
+        # Perform section-specific linting
+        state = "module"
+        process_lines = []
+        for l in lines:
+            if l.startswith("process") and state == "module":
+                state = "process"
+            if re.search("input\s*:", l) and state == "process":
+                state = "input"
+                continue
+            if re.search("output\s*:", l) and state == "input":
+                state = "output"
+                continue
+            if re.search("script\s*:", l) and state == "output":
+                state = "script"
+                continue
+
+            # Perform state-specific linting checks
+            if state == "process" and not self._is_empty(l):
+                process_lines.append(l)
+            if state == "input" and not self._is_empty(l):
+                inputs += self._parse_input(l)
+            if state == "output" and not self._is_empty(l):
+                outputs += self._parse_output(l)
+                outputs = list(set(outputs))  # remove duplicate 'meta's
+
+        # Check the process defintions
+        if self.check_process_section(process_lines):
+            self.passed.append("Matching build versions in {}".format(file))
+        else:
+            self.failed.append("Build versions are not matching: {}".format(file))
+
+        # Check whether 'meta' is emitted when given as input
+        if "meta" in inputs:
+            if "meta" in outputs:
+                self.passed.append("'meta' emitted in {}".format(file))
+            else:
+                self.failed.append("'meta' given as input but not emitted in {}".format(file))
+
+        # Check that a software version is emitted
+        if "version" in outputs:
+            self.passed.append("Module emits software version: {}".format(file))
+        else:
+            self.failed.append("Module doesn't emit  software version {}".format(file))
+
+        return inputs, outputs
+
+    def check_process_section(self, lines):
+        """
+        Lint the section of a module between the process definition
+        and the 'input:' definition
+        Specifically checks for correct software versions
+        and containers
+        """
+        # Checks that build numbers of bioconda, singularity and docker container are matching
+        build_id = "build"
+        singularity_tag = "singularity"
+        docker_tag = "docker"
+
+        for l in lines:
+            if re.search("bioconda::", l):
+                bioconda = l.split()
+                bioconda = [b for b in bioconda if "bioconda" in b][0]
+                build_id = bioconda.split("::")[1].replace('"', "").replace("'", "").split("=")[-1].strip()
```

Comment:
```suggestion
                build_id = bioconda.split("::")[1].strip("'\"").split("=")[-1].strip()
```

Reply:
Could be nice to do some refactoring of the code that we have that checks `environment.yml` and bring in what we already have where it checks that the conda package version can be found and is the latest available.

Reply:
Ou if that works that's awesome, I was super unhappy with these double-replace things.
Yes that was still on my todo-list, to check for the conda version.

This is also the section of the code that I like the least, just a lot of string manipulation which I'm afraid of breaking some day .... but couldn't think of anything better yet because there isn't too much structure to that part of modules.